toolchain/parse/state.def

```diff
@@ -741,6 +746,40 @@ CARBON_PARSE_STATE(NamespaceFinish)
 //   (state done)
 CARBON_PARSE_STATE(Package)
 
+// Handles `require` after the introducer.
+//
+// require ...
+//        ^
+//   1. Expr
+//   2. Impls
+//   3. RequireFinish
+//
+CARBON_PARSE_STATE(Require)
+
+// Handles the processing of `impls` in a `require` declaration after the type
+// expression.
+// impls ...
+// ^~~~~
+//
+//   1. Expr
+//   2. ImplsFinish
+//
```

Comment:
The syntax you're documenting here is for the error case, but you have it for the expression case. Let's document both paths.

Also note this suggestion is fixing extra blank lines.

```suggestion
// expr impls ...
//      ^~~~~
//   1. Expr
//   2. ImplsFinish
//
// expr ???
//     ^
//   (state done)
```